### ISSUE TITLE
Triangular Artifacts When Rendering ShapeNet with No Texture

### ISSUE THREAD

**Issue by b7leung** [facebookresearch/pytorch3d#313]
## 🐛 Bugs / Unexpected behaviors

When I render ShapeNet with no texture (left), there seems to be a lot of triangular artifacts. It should look more like when I open the obj with meshlab (right). I'm not very experienced with rendering, so I'm not sure if this is a bug or I have some settings set incorrectly.

![pytorch_3d](https://user-images.githubusercontent.com/20059131/90194612-88dba280-dd7c-11ea-8b2d-f84e49f0c47f.png)![meshlab](https://user-images.githubusercontent.com/20059131/90194616-89743900-dd7c-11ea-9587-a4b936dcd448.png)

## Instructions To Reproduce the Issue:

The obj file above is from ShapeNet V1 chair class, instance 1a74a83fa6d24b3cacd67ce2c72c02e. But the issue is the same for other obj files from ShapeNet as well. The model is:

[model.zip](https://github.com/facebookresearch/pytorch3d/files/5071848/model.zip)

The code I'm using to load the mesh without texture is based on #51. I'm also using the latest nightly version of pytorch3d.

```
import torch
import matplotlib.pyplot as plt
from pytorch3d.io import load_objs_as_meshes
from pytorch3d.io import load_obj
from pytorch3d.structures import Meshes
from pytorch3d.renderer import Textures
from pytorch3d.renderer import (
    look_at_view_transform,
    OpenGLPerspectiveCameras, 
    PointLights, 
    DirectionalLights, 
    Materials, 
    RasterizationSettings, 
    MeshRenderer, 
    MeshRasterizer,  
    SoftPhongShader,
    HardPhongShader,
    SoftSilhouetteShader,
    BlendParams
)

# based on https://github.com/facebookresearch/pytorch3d/issues/51
def load_untextured_mesh(mesh_path, device):
    mesh = load_objs_as_meshes([mesh_path], device=device)
    verts, faces_idx, _ = load_obj(mesh_path)
    faces = faces_idx.verts_idx
    # Initialize each vertex to be white in color
    verts_rgb = torch.ones_like(verts)[None] # (1, V, 3)
    textures = Textures(verts_rgb=verts_rgb.to(device))
    mesh_no_texture = Meshes(
        verts=[verts.to(device)],
        faces=[faces.to(device)],
        textures=textures
        )
    return mesh_no_texture

def render_mesh(mesh, R, T, device, img_size=512):
    cameras = OpenGLPerspectiveCameras(device=device, R=R, T=T)
    raster_settings = RasterizationSettings(
        image_size=img_size, 
        blur_radius=0.0, 
        faces_per_pixel=1, 
    )
    lights = PointLights(device=device, location=[[0.0, 5.0, -10.0]])
    renderer = MeshRenderer(
        rasterizer=MeshRasterizer(
            cameras=cameras, 
            raster_settings=raster_settings
        ),
        shader=SoftPhongShader(
            device=device, 
            cameras=cameras,
            lights=lights
        )
    )
    rendered_images = renderer(mesh, cameras=cameras)
    return rendered_images

obj_model_path = "data/example_shapenet/03001627/1a74a83fa6d24b3cacd67ce2c72c02e/model.obj"
device = torch.device("cuda:1")

with torch.no_grad():
    mesh = load_untextured_mesh(obj_model_path, device)
    R, T = look_at_view_transform(1, 30, 45) 
    render = render_mesh(mesh, R, T, device)
plt.imshow(render[0, ..., :3].cpu().numpy())

```

**Comment by nikhilaravi** [comment]
Hi @b7leung this is to do with the lighting! I am looking into this issue. You can turn off the lighting temporarily by setting

```
lights = PointLights(
    device=device, 
    location=[[0.0, 5.0, -10.0]], 
    diffuse_color=((0, 0, 0),),
    specular_color=((0, 0, 0),),
)
``` 

This should get rid of the artifacts.

**Comment by b7leung** [comment]
Thanks for the response, and for looking into this issue! I'm writing a paper on single view reconstruction and pytorch3d has been a lifesaver. 

Anyways, it seems like turning off the lighting does get rid of the artifacts, but also causes the render to essentially become a silhouette. If there are any other workarounds/fixes, that would be great.

**Comment by nikhilaravi** [comment]
That's so great to hear, glad you are finding PyTorch3D useful. I will reply back here when I find a fix for the lighting!

**Comment by gkioxari** [comment]
I also noticed that this could be an artifact of the position of the lighting? @b7leung have you tried putting the lighting in a different location?

**Comment by b7leung** [comment]
> I also noticed that this could be an artifact of the position of the lighting. @b7leung have you tried putting the lighting in a different location?

Changing the light position seems to cause the artifacts to show up differently. For example, the following are rendered with the light at position [1.0, 1.0, 1.0], [10,10,1], [-10,-10,-10], and [100,100,100] respectively.

![1_1_1](https://user-images.githubusercontent.com/20059131/90307705-b3695080-de8d-11ea-8b69-caa8a111ecbe.png)![10_10_1](https://user-images.githubusercontent.com/20059131/90307706-b5cbaa80-de8d-11ea-9b21-391b1e130dbc.png)
![neg_10_10_10](https://user-images.githubusercontent.com/20059131/90307708-b7956e00-de8d-11ea-961d-37614c2e7ca3.png)![100_100_100](https://user-images.githubusercontent.com/20059131/90307709-b95f3180-de8d-11ea-89cc-6c12ffd16425.png)

**Comment by gkioxari** [comment]
Hi @b7leung 

What we care about here is if there is a bug in the PyTorch3D renderer. And that is not clear to me. 

It is indeed the case that viewers (forward renderers) do some smoothing (these are usually by default choices in the viewers and can either take the form of applying non flat shading to the output or other such manipulations). In this case, non flat shading could explain the discrepancies in the two rendered outputs, i.e. from PyTorch3D and from your viewer. These smoothing operations might be standard for viewers but are not appropriate for learning libraries as they change the output and lead to incorrect gradients. 

If you believe there is a but though, we should definitely fix it but we would have to first pin it down and isolate it.

**Comment by gkioxari** [comment]
Hi @b7leung 
I looked into this and the discrepancies in the outputs are with regards to the shader and in particular with how face normals are computed. For example, here is MeshLab's output when you choose `Vert` shading instead of `Face` (which is the default)

![Screen Shot 2020-08-23 at 8 35 35 PM](https://user-images.githubusercontent.com/4369065/91001767-7e41ba00-e581-11ea-96f2-f296692a03b7.png)

This is not a bug but, as I said, has to do with how face normals are computed, basically either by interpolating them from vertex normals or by passing them as inputs. Phong shading assumes a smoothly varying surface normal vector and a well designed CAD model should have those and lead to smooth interpolated face normals from vertex normals. However, a badly designed CAD model, as is the case with the model here, will lead to different face normals when they are interpolated from vertex normals (as done in `Vert` and also in Phong shading). 

I think this is an interesting case and raises the question of whether we want to support face normals being used directly or instead store vertex normals and interpolate face normals from them. We do the latter.

**Comment by b7leung** [comment]
Yeah, that really helps. Thank you!
